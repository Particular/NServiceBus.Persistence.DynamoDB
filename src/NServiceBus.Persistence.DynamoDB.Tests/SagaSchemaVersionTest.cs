namespace NServiceBus.Persistence.DynamoDB.Tests
{
    using System;
    using System.Linq;
    using System.Text.Json;
    using System.Threading.Tasks;
    using Extensibility;
    using NUnit.Framework;
    using Particular.Approvals;
    using Sagas;
    using Testing;

    [TestFixture]
    public class SagaSchemaVersionTest
    {
        [Test]
        public async Task Should_update_schema_version_on_schema_changes()
        {
            var sagaPersister = new SagaPersister(new MockDynamoDBClient(), new SagaPersistenceConfiguration(), "SchemaVersionTest");

            var sagaData = new TestSagaData()
            {
                CorrelationProperty = "CorrelationPropertyValue",
                Id = new Guid("FFC8A2FD-0335-47C8-A29D-9EEA6C8445D8"),
                OriginalMessageId = "OriginalMessageIdValue",
                Originator = "OriginatorValue"
            };

            var testableSession = new TestableDynamoDBSynchronizedStorageSession();
            await sagaPersister.Save(sagaData,
                new SagaCorrelationProperty(nameof(TestSagaData.CorrelationProperty), sagaData.CorrelationProperty),
                testableSession, new ContextBag());

            // !!! IMPORTANT !!!
            // This test should help to detect data schema changes.
            // When this test fails, make sure to update the saga data schema version in the metadata.
            Approver.Verify(
                JsonSerializer.Serialize(testableSession.TransactWriteItems.Single().Put.Item,
                    new JsonSerializerOptions { WriteIndented = true }));
        }

        class TestSagaData : ContainSagaData
        {
            public string CorrelationProperty { get; set; }
        }
    }
}