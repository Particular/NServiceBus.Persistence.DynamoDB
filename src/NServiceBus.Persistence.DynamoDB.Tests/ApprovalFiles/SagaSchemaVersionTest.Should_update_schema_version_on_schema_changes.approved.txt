{
  "CorrelationProperty": {
    "B": null,
    "BOOL": false,
    "IsBOOLSet": false,
    "BS": [],
    "L": [],
    "IsLSet": false,
    "M": {},
    "IsMSet": false,
    "N": null,
    "NS": [],
    "NULL": false,
    "S": "CorrelationPropertyValue",
    "SS": []
  },
  "Id": {
    "B": null,
    "BOOL": false,
    "IsBOOLSet": false,
    "BS": [],
    "L": [],
    "IsLSet": false,
    "M": {},
    "IsMSet": false,
    "N": null,
    "NS": [],
    "NULL": false,
    "S": "ffc8a2fd-0335-47c8-a29d-9eea6c8445d8",
    "SS": []
  },
  "Originator": {
    "B": null,
    "BOOL": false,
    "IsBOOLSet": false,
    "BS": [],
    "L": [],
    "IsLSet": false,
    "M": {},
    "IsMSet": false,
    "N": null,
    "NS": [],
    "NULL": false,
    "S": "OriginatorValue",
    "SS": []
  },
  "OriginalMessageId": {
    "B": null,
    "BOOL": false,
    "IsBOOLSet": false,
    "BS": [],
    "L": [],
    "IsLSet": false,
    "M": {},
    "IsMSet": false,
    "N": null,
    "NS": [],
    "NULL": false,
    "S": "OriginalMessageIdValue",
    "SS": []
  },
  "PK": {
    "B": null,
    "BOOL": false,
    "IsBOOLSet": false,
    "BS": [],
    "L": [],
    "IsLSet": false,
    "M": {},
    "IsMSet": false,
    "N": null,
    "NS": [],
    "NULL": false,
    "S": "SAGA#ffc8a2fd-0335-47c8-a29d-9eea6c8445d8",
    "SS": []
  },
  "SK": {
    "B": null,
    "BOOL": false,
    "IsBOOLSet": false,
    "BS": [],
    "L": [],
    "IsLSet": false,
    "M": {},
    "IsMSet": false,
    "N": null,
    "NS": [],
    "NULL": false,
    "S": "SAGA#ffc8a2fd-0335-47c8-a29d-9eea6c8445d8",
    "SS": []
  },
  "NSERVICEBUS_METADATA": {
    "B": null,
    "BOOL": false,
    "IsBOOLSet": false,
    "BS": [],
    "L": [],
    "IsLSet": false,
    "M": {
      "VERSION": {
        "B": null,
        "BOOL": false,
        "IsBOOLSet": false,
        "BS": [],
        "L": [],
        "IsLSet": false,
        "M": {},
        "IsMSet": false,
        "N": "0",
        "NS": [],
        "NULL": false,
        "S": null,
        "SS": []
      },
      "TYPE": {
        "B": null,
        "BOOL": false,
        "IsBOOLSet": false,
        "BS": [],
        "L": [],
        "IsLSet": false,
        "M": {},
        "IsMSet": false,
        "N": null,
        "NS": [],
        "NULL": false,
        "S": "NServiceBus.Persistence.DynamoDB.Tests.SagaSchemaVersionTest\u002BTestSagaData",
        "SS": []
      },
      "SCHEMA_VERSION": {
        "B": null,
        "BOOL": false,
        "IsBOOLSet": false,
        "BS": [],
        "L": [],
        "IsLSet": false,
        "M": {},
        "IsMSet": false,
        "N": null,
        "NS": [],
        "NULL": false,
        "S": "1.0.0",
        "SS": []
      }
    },
    "IsMSet": true,
    "N": null,
    "NS": [],
    "NULL": false,
    "S": null,
    "SS": []
  },
  "LEASE_TIMEOUT": {
    "B": null,
    "BOOL": false,
    "IsBOOLSet": false,
    "BS": [],
    "L": [],
    "IsLSet": false,
    "M": {},
    "IsMSet": false,
    "N": "-1",
    "NS": [],
    "NULL": false,
    "S": null,
    "SS": []
  }
}